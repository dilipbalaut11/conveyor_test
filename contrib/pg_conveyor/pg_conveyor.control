# pg_conveyor test extension
comment = 'test conveyor'
default_version = '1.0'
module_pathname = '$libdir/pg_conveyor'
relocatable = true
