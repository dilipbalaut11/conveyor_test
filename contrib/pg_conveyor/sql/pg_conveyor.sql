CREATE EXTENSION pg_conveyor;

CREATE TABLE test(a int);

SELECT pg_conveyor_init('test'::regclass::oid, 4);
SELECT pg_conveyor_insert('test'::regclass::oid, 'test_data');
SELECT pg_conveyor_read('test'::regclass::oid, 0);

--CASE1
do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..1000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;

-- read from some random blocks
SELECT pg_conveyor_read('test'::regclass::oid, 100);
SELECT pg_conveyor_read('test'::regclass::oid, 800);

--CASE2
do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..5000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i+1000;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
SELECT pg_conveyor_read('test'::regclass::oid, 4000);
SELECT pg_conveyor_read('test'::regclass::oid, 3000);

--CASE3
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..50000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;


--CASE4--(vacuum is failing)
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..5000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;

SELECT pg_conveyor_truncate('test'::regclass::oid, 3000);
SELECT pg_conveyor_vacuum('test'::regclass::oid);

--CASE5
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..50000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;

--CASE6 (multi truncate single vacuum)
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..1000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;

--CASE7
SELECT pg_conveyor_truncate('test'::regclass::oid, 500);
do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..1000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;

SELECT pg_conveyor_truncate('test'::regclass::oid, 1800);
SELECT pg_conveyor_vacuum('test'::regclass::oid);

--CASE8
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..5000000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
