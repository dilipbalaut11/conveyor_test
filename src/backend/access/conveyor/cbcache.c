/*-------------------------------------------------------------------------
 *
 * cbcache.c
 *	  Conveyor belt index segment location cache.
 *
 * The conveyor belt metapage stores the segment numbers of the oldest and
 * newest index segments that currently exist, but the location of any
 * other index segment can only be discovered by reading the first page
 * of some index segment whose position is known and extracting from it
 * the segment number of the next index segment. That's potentially
 * expensive, especially if we have to traverse a whole bunch of index
 * segments to figure out the location of the one we need, so we maintain
 * a cache.
 *
 * The cache maps the oldest logical page number covered by an index
 * segment to the segment number where that index segment is located.
 * If older index segments are removed, the corresponding mappings become
 * obsolete, but nobody should be accessing those pages anyway. Still,
 * we're careful to purge such mappings to avoid wasting memory.
 *
 * If an index segment is moved, we invalidate the entire cache. This
 * is expected to be fairly rare, as it should only happen if someone is
 * trying to reduce the on-disk footprint of the conveyor belt. Moreover,
 * if someone is doing that, it is likely that multiple index segments
 * will be moved in relatively quick succession, so it's not clear that
 * a more granular invalidation strategy would help anything.
 *
 * Copyright (c) 2016-2021, PostgreSQL Global Development Group
 *
 * src/backend/access/conveyor/cbcache.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/cbcache.h"
#include "common/hashfn.h"

typedef struct cb_iseg_entry
{
	CBPageNo	index_segment_start;
	CBSegNo		segno;
	char		status;
} cb_iseg_entry;

#define SH_PREFIX cb_iseg
#define SH_ELEMENT_TYPE cb_iseg_entry
#define SH_KEY_TYPE CBPageNo
#define SH_KEY index_segment_start
#define SH_HASH_KEY(tb, key) tag_hash(&key, sizeof(CBPageNo))
#define SH_EQUAL(tb, a, b) a == b
#define SH_SCOPE static inline
#define SH_DECLARE
#define SH_DEFINE
#include "lib/simplehash.h"

struct CBCache
{
	uint64		index_segments_moved;
	CBPageNo	oldest_possible_start;
	cb_iseg_hash *iseg;
};

/*
 * Create a new cache.
 */
CBCache *
cb_cache_create(MemoryContext mcxt, uint64 index_segments_moved)
{
	CBCache    *cache = MemoryContextAlloc(mcxt, sizeof(CBCache));

	cache->index_segments_moved = index_segments_moved;
	cache->oldest_possible_start = 0;
	cache->iseg = cb_iseg_create(mcxt, 128, NULL);

	return cache;
}

/*
 * Invalidate cache entries as required.
 *
 * If index_segments_moved has changed, throw away everything we think we
 * know. Otherwise, if index_start has advanced, throw away any entries that
 * precede the new start.
 */
void
cb_cache_invalidate(CBCache *cache, CBPageNo index_start,
					uint64 index_segments_moved)
{
	if (index_segments_moved != cache->index_segments_moved)
	{
		cb_iseg_reset(cache->iseg);
		cache->index_segments_moved = index_segments_moved;
	}
	else if (index_start > cache->oldest_possible_start)
	{
		cb_iseg_iterator it;
		cb_iseg_entry *entry;

		cb_iseg_start_iterate(cache->iseg, &it);
		while ((entry = cb_iseg_iterate(cache->iseg, &it)) != NULL)
			if (entry->index_segment_start < index_start)
				cb_iseg_delete_item(cache->iseg, entry);
	}
}

/*
 * Search the cache for an index segment number, given the first logical page
 * number covered by that index segment.
 *
 * It is the caller's responsibility to make sure that pageno is the first
 * logical page number covered by some index segment, rather than any random
 * page number whose index entry might be anywhere in the segment. We don't
 * have enough information here to verify this, and just trust that the caller
 * knows what they are doing.
 */
CBSegNo
cb_cache_lookup(CBCache *cache, CBPageNo pageno)
{
	cb_iseg_entry *entry;

	entry = cb_iseg_lookup(cache->iseg, pageno);
	return entry != NULL ? entry->segno : CB_INVALID_SEGMENT;
}

/*
 * Search the cache for an index segment that precedes the one for which we
 * are searching by as little as possible.
 *
 * As with cb_cache_lookup, pageno should be the first logical page of the
 * index segment in which the caller is interested, although unlike that
 * function, this function would still work correctly if it were an arbitrary
 * page number, at least as presently implemented.
 *
 * If no segment with a starting segment number preceding pageno is found
 * in cache, the return value is CB_INVALID_SEGMENT and *index_segment_start
 * is set to CB_INVALID_LOGICAL_PAGE. Otherwise, the return value is the
 * segment number we found and *index_segment_start is set to the starting
 * logical page number of that segment.
 */
CBSegNo
cb_cache_fuzzy_lookup(CBCache *cache, CBPageNo pageno,
					  CBPageNo *index_segment_start)
{
	cb_iseg_iterator it;
	cb_iseg_entry *current;
	cb_iseg_entry *best = NULL;

	cb_iseg_start_iterate(cache->iseg, &it);
	while ((current = cb_iseg_iterate(cache->iseg, &it)) != NULL)
	{
		if (current->index_segment_start > pageno)
			continue;
		if (best == NULL ||
			best->index_segment_start < current->index_segment_start)
			best = current;
	}

	if (best == NULL)
	{
		*index_segment_start = CB_INVALID_LOGICAL_PAGE;
		return CB_INVALID_SEGMENT;
	}

	*index_segment_start = best->index_segment_start;
	return best->segno;
}

/*
 * Insert a cache entry.
 *
 * As in cb_cache_lookup, it's critical that index_segment_start is the first
 * logical page number covered by the index segment.
 */
void
cb_cache_insert(CBCache *cache, CBSegNo segno, CBPageNo index_segment_start)
{
	cb_iseg_entry *entry;
	bool		found;

	entry = cb_iseg_insert(cache->iseg, index_segment_start, &found);
	Assert(!found);
	entry->segno = segno;

	Assert(index_segment_start >= cache->oldest_possible_start);
}
