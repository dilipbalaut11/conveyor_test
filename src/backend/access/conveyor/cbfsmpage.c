/*-------------------------------------------------------------------------
 *
 * cbfsmpage.c
 *	  APIs for accessing conveyor belt FSM pages.
 *
 * Similar to cbmetapage.c, this file abstracts accesses to conveyor
 * belt FSM pages, and should be the only backend code that understands
 * their internal structure.
 *
 * Copyright (c) 2016-2021, PostgreSQL Global Development Group
 *
 * src/backend/access/conveyor/cbfsmpage.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/cbfsmpage.h"
#include "access/cbfsmpage_format.h"
#include "access/cbmetapage.h"

static CBFSMPageData *cb_fsmpage_get_special(Page page);

/*
 * Initialize FSM page.
 *
 * Returns the first segment number that will be covered by the new page.
 */
CBSegNo
cb_fsmpage_initialize(Page page, BlockNumber blkno, uint16 pages_per_segment)
{
	CBFSMPageData *fsmp;
	BlockNumber	first_fsm_block = cb_first_fsm_block(pages_per_segment);
	unsigned	fsm_block_spacing = cb_fsm_block_spacing(pages_per_segment);

	/* Sanity checks. */
	Assert(blkno >= first_fsm_block);
	Assert((blkno - first_fsm_block) % fsm_block_spacing == 0);

	/* Initialize page. PageInit will zero the payload bits for us. */
	PageInit(page, BLCKSZ, sizeof(CBFSMPageData));
	fsmp = (CBFSMPageData *) PageGetSpecialPointer(page);
	fsmp->cbfsm_magic = CB_FSMPAGE_MAGIC;
	fsmp->cbfsm_start = cb_first_segment_for_fsm_page(blkno, pages_per_segment);

	return fsmp->cbfsm_start;
}

/*
 * Get the allocation status of a segment from an FSM page.
 */
bool
cb_fsmpage_get_fsm_bit(Page page, CBSegNo segno)
{
	CBFSMPageData *fsmp = cb_fsmpage_get_special(page);
	uint8		byte;
	uint8		mask;
	uint32		bitno;

	if (segno < fsmp->cbfsm_start ||
		segno >= fsmp->cbfsm_start + CB_FSM_SEGMENTS_PER_FSMPAGE)
		elog(ERROR, "segment %u out of range for fsm page starting at segment %u",
			 segno, fsmp->cbfsm_start);

	bitno = segno - fsmp->cbfsm_start;
	byte = fsmp->cbfsm_state[bitno / BITS_PER_BYTE];
	mask = 1 << (bitno % BITS_PER_BYTE);
	return (segno & mask) != 0;
}

/*
 * Set the allocation status of a segment in an FSM page.
 *
 * new_state should be true if the bit is currently clear and should be set,
 * and false if the bit is currently set and should be cleared. Don't call
 * this unless you know that the bit actually needs to be changed.
 */
void
cb_fsmpage_set_fsm_bit(Page page, CBSegNo segno, bool new_state)
{
	CBFSMPageData *fsmp = cb_fsmpage_get_special(page);
	uint8	   *byte;
	uint8		mask;
	uint8		old_state;
	uint32		bitno;

	if (segno < fsmp->cbfsm_start ||
		segno >= fsmp->cbfsm_start + CB_FSM_SEGMENTS_PER_FSMPAGE)
		elog(ERROR, "segment %u out of range for fsm page starting at segment %u",
			 segno, fsmp->cbfsm_start);

	bitno = segno - fsmp->cbfsm_start;
	byte = &fsmp->cbfsm_state[bitno / BITS_PER_BYTE];
	mask = 1 << (segno % BITS_PER_BYTE);
	old_state = (*byte & mask) != 0;

	if (old_state == new_state)
		elog(ERROR, "fsm bit for segment %u already has value %d",
			 segno, old_state ? 1 : 0);

	if (new_state)
		*byte |= mask;
	else
		*byte &= ~mask;
}

/*
 * Returns the lowest unused segment number covered by the supplied FSM page,
 * or CB_INVALID_SEGMENT if none.
 */
CBSegNo
cbfsmpage_find_free_segment(Page page)
{
	CBFSMPageData *fsmp = cb_fsmpage_get_special(page);
	unsigned	i;

	StaticAssertStmt(CB_FSMPAGE_FREESPACE_BYTES % sizeof(uint64) == 0,
					 "CB_FSMPAGE_FREESPACE_BYTES should be a multiple of 8");

	for (i = 0; i < CB_FSMPAGE_FREESPACE_BYTES; i += sizeof(uint64))
	{
		uint64	word = * (uint64 *) &fsmp->cbfsm_state[i];

		if (word != PG_UINT64_MAX)
		{
			uint64		flipword = ~word;
			int			b = fls((int) flipword);

			if (b == 0)
				b = 32 + fls((int) (flipword >> 32));

			Assert(b >= 1 && b <= 64);
			return fsmp->cbfsm_start + (i * BITS_PER_BYTE) + (b - 1);
		}
	}

	return CB_INVALID_SEGMENT;
}

/*
 * Given a page that is known to be a conveyor belt free space map page,
 * return a pointer to the CBFSMPageData, after checking the magic number.
 */
static CBFSMPageData *
cb_fsmpage_get_special(Page page)
{
	CBFSMPageData *fsmp = (CBFSMPageData *) PageGetSpecialPointer(page);

	if (fsmp->cbfsm_magic != CB_FSMPAGE_MAGIC)
		elog(ERROR, "bad magic number in conveyor belt fsm page: %08X",
			 fsmp->cbfsm_magic);

	return fsmp;
}
