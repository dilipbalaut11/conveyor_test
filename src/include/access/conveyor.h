/*-------------------------------------------------------------------------
 *
 * conveyor.h
 *	  Public API for conveyor belt storage.
 *
 * See src/backend/access/conveyor/README for a general overview of
 * conveyor belt storage.
 *
 * Copyright (c) 2016-2021, PostgreSQL Global Development Group
 *
 * src/include/access/conveyor.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef CONVEYOR_H
#define CONVEYOR_H

#include "access/cbdefs.h"
#include "common/relpath.h"
#include "utils/relcache.h"
#include "storage/bufmgr.h"

struct ConveyorBelt;
typedef struct ConveyorBelt ConveyorBelt;

/* Routines to create a new conveyor belt, or open an existing one. */
extern ConveyorBelt *ConveyorBeltInitialize(Relation rel,
											ForkNumber fork,
											uint16 pages_per_segment,
											MemoryContext mcxt);
extern ConveyorBelt *ConveyorBeltOpen(Relation rel,
									  ForkNumber fork,
									  MemoryContext mcxt);

/* Routines to inserting new data into a conveyor belt. */
extern Buffer ConveyorBeltGetNewPage(ConveyorBelt *cb, CBPageNo *pageno);
extern void ConveyorBeltPerformInsert(ConveyorBelt *cb, Buffer buffer,
									  bool page_std);
extern void ConveyorBeltCleanupInsert(ConveyorBelt *cb, Buffer buffer);

/* Routines for reading data from a conveyor belt. */
extern Buffer ConveyorBeltReadBuffer(ConveyorBelt *cb, CBPageNo pageno,
									 int mode,
									 BufferAccessStrategy strategy);
extern void ConveyorBeltGetBounds(ConveyorBelt *cb,
								  CBPageNo *oldest_logical_page,
								  CBPageNo *next_logical_page);

/* Routines for removing old data from a conveyor belt. */
extern void ConveyorBeltLogicalTruncate(ConveyorBelt *cb,
										CBPageNo oldest_keeper);
extern void ConveyorBeltVacuum(ConveyorBelt *cb);
extern void ConveyorBeltPhysicalTruncate(ConveyorBelt *cb);
extern void ConveyorBeltCompact(ConveyorBelt *cb);
extern ConveyorBelt *ConveyorBeltRewrite(ConveyorBelt *cb,
										 Relation rel,
										 ForkNumber fork,
										 MemoryContext mcxt);

#endif							/* CONVEYOR_H */
