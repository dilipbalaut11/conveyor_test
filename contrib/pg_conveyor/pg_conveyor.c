/*-------------------------------------------------------------------------
 *
 * pg_conveyor.c
 *	  provide test api to create/write/read conveyor belt infrastructure
 *
 * Copyright (c) 2016-2021, PostgreSQL Global Development Group
 *
 * contrib/pg_conveyor/pg_conveyor.c
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/conveyor.h"
#include "access/relation.h"
#include "fmgr.h"
#include "miscadmin.h"
#include "storage/bufpage.h"
#include "storage/smgr.h"
#include "utils/builtins.h"
#include "utils/rel.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(pg_conveyor_init);
PG_FUNCTION_INFO_V1(pg_conveyor_insert);
PG_FUNCTION_INFO_V1(pg_conveyor_read);


/*
 * Initialize a new conveyor belt for input relid.
 */
Datum
pg_conveyor_init(PG_FUNCTION_ARGS)
{
	Oid	relid = PG_GETARG_OID(0);
	SMgrRelation	reln;
	Relation		rel;

	rel = relation_open(relid, AccessShareLock);

	/* Open the relation at smgr level. */
	reln = RelationGetSmgr(rel);

	/*
	 * If the dead_tid fork doesn't exist then create it and initialize the
	 * conveyor belt, otherwise just open the conveyor belt.
	 */
	if (!smgrexists(reln, DEADTID_FORKNUM))
	{
		smgrcreate(reln, DEADTID_FORKNUM, false);
		ConveyorBeltInitialize(rel, DEADTID_FORKNUM, 4, CurrentMemoryContext);
	}

	relation_close(rel, AccessShareLock);

	/* Nothing to return. */
	PG_RETURN_VOID();
}

/*
 * insert input buffer data into the conveyor belt.
 */
Datum
pg_conveyor_insert(PG_FUNCTION_ARGS)
{
	Oid	relid = PG_GETARG_OID(0);
	char   *data = text_to_cstring(PG_GETARG_TEXT_PP(1));
	Relation	rel;
	ConveyorBelt   *cb;
	CBPageNo		pageno;
	Buffer			buffer;
	PageHeader		phdr;
	Page			page;
	char		   *pagedata;

	rel = relation_open(relid, AccessExclusiveLock);

	/* Open the conveyor belt. */
	cb = ConveyorBeltOpen(rel, DEADTID_FORKNUM, CurrentMemoryContext);

	buffer = ConveyorBeltGetNewPage(cb, &pageno);
	page = BufferGetPage(buffer);
	pagedata = PageGetContents(page);
	PageInit(page, BLCKSZ, 0);

	phdr = (PageHeader) page;

	START_CRIT_SECTION();
	memcpy(pagedata, data, strlen(data));
	phdr->pd_lower += strlen(data);
	ConveyorBeltPerformInsert(cb, buffer);
	END_CRIT_SECTION();

	ConveyorBeltCleanupInsert(cb, buffer);

	relation_close(rel, AccessExclusiveLock);

	/* Nothing to return. */
	PG_RETURN_VOID();
}

/*
 * read data page data from the relation's conveyor belt.
 */
Datum
pg_conveyor_read(PG_FUNCTION_ARGS)
{
	Oid		relid = PG_GETARG_OID(0);
	Oid		pageno = PG_GETARG_INT32(1);
	Relation		rel;
	ConveyorBelt   *cb;
	Buffer			buffer;
	char			pagedata[BLCKSZ];

	rel = relation_open(relid, AccessShareLock);
	cb = ConveyorBeltOpen(rel, DEADTID_FORKNUM, CurrentMemoryContext);

	buffer = ConveyorBeltReadBuffer(cb, pageno, BUFFER_LOCK_SHARE, NULL);
	if (BufferIsInvalid(buffer))
		elog(ERROR, "could not read data");

	memcpy(pagedata, BufferGetPage(buffer), BLCKSZ);
	UnlockReleaseBuffer(buffer);

	relation_close(rel, AccessShareLock);

	PG_RETURN_DATUM(CStringGetTextDatum((char *) PageGetContents((char *) pagedata)));
}
