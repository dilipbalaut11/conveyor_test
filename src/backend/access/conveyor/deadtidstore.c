/*-------------------------------------------------------------------------
 *
 * deadtidstore.c
 *	  Store and fetch deadtids.
 *
 * Create dead tid fork and manage storing and retriving deadtids using
 * conveyor belt infrastructure.
 *
 * Copyright (c) 2016-2021, PostgreSQL Global Development Group
 *
 * src/backend/access/conveyor/deadtidstore.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/conveyor.h"
#include "access/deadtidstore.h"
#include "miscadmin.h"
#include "storage/smgr.h"
#include "utils/rel.h"

#define DTS_PAGES_PER_SEGMENT	4
#define DTS_TID_PERPAGE \
	(BLCKSZ - MAXALIGN(SizeOfPageHeaderData)) / sizeof(ItemPointerData)
#define DTS_PAGE_TID_COUNT(page) \
	(((PageHeader) (page))->pd_lower -  MAXALIGN(SizeOfPageHeaderData)) / sizeof(ItemPointerData)

/*
 *	DTS_AppendTid() -- flush out dead TIDs to the dead tid store
 *
 *  *pageno - output logical pageno from where it started dumping the deadtids.
 */
void
DTS_AppendTid(Relation rel, int ntids, ItemPointerData *deadtids)
{
	ConveyorBelt   *cb;
	SMgrRelation	reln;

	/* Open the relation at smgr level. */
	reln = RelationGetSmgr(rel);

	/*
	 * If the dead_tid fork doesn't exist then create it and initialize the
	 * conveyor belt, otherwise just open the conveyor belt.
	 */
	if (!smgrexists(reln, DEADTID_FORKNUM))
	{
		smgrcreate(reln, DEADTID_FORKNUM, false);
		cb = ConveyorBeltInitialize(rel,
									DEADTID_FORKNUM,
									DTS_PAGES_PER_SEGMENT,
									CurrentMemoryContext);
	}
	else
		cb = ConveyorBeltOpen(rel, DEADTID_FORKNUM, CurrentMemoryContext);

	/* Loop until we store all dead tids in the conveyor belt. */
	while(ntids > 0)
	{
		Buffer	buffer;
		Page	page;
		char   *pagedata;
		int		count;
		PageHeader	phdr;
		CBPageNo	pageno;

		/* Get a new page from the ConveyorBelt. */
		buffer = ConveyorBeltGetNewPage(cb, &pageno);
		page = BufferGetPage(buffer);
		PageInit(page, BLCKSZ, 0);
		phdr = (PageHeader) page;
		pagedata = PageGetContents(page);

		/* Compute how many dead tids we can store into this page. */
		count =  Min(DTS_TID_PERPAGE, ntids);

		START_CRIT_SECTION();
		/*
		 * Copy the data into the page and set the pd_lower.  We are just
		 * copying ItemPointerData array so we don't need any alignment which
		 * should be SHORTALIGN and sizeof(ItemPointerData) will take care of
		 * that so we don't need any other alignment.
		 */
		memcpy(pagedata, deadtids,
			   sizeof(ItemPointerData) * count);
		phdr->pd_lower += sizeof(ItemPointerData) * count;
		ConveyorBeltPerformInsert(cb, buffer);

		END_CRIT_SECTION();

		ConveyorBeltCleanupInsert(cb, buffer);

		/* Adjust deadtid pointer and the remaining tid count. */
		deadtids += count;
		ntids -= count;
	}
}

/*
 * DTS_ReadDeadTids - Read dead tids from the deat tid store
 *
 * start_pageno - Logical pageno from where to start reading the dead tids
 * deadtids - Pre allocated array for holding the dead tids, the allocated size
 * is sufficient to hold maxtids.
 * *last_pageread - Store last page number we have read dead tids from.
 *
 * Returns number of dead tids actually read from the storage.
 */
int
DTS_ReadDeadTids(Relation rel, CBPageNo start_pageno, int maxtids,
				 ItemPointerData *deadtids, CBPageNo *last_pageread)
{
	SMgrRelation	reln;
	ConveyorBelt   *cb;
	Page			page;
	Buffer			buffer;
	CBPageNo		oldest_pageno;
	CBPageNo		next_pageno;
	int				count;
	int				ndeadtids = 0;
	int				prevpageno = start_pageno;

	/* Open the relation at smgr level. */
	reln = RelationGetSmgr(rel);

	/* The deadtid fork must exists. */
	Assert(smgrexists(reln, DEADTID_FORKNUM));

	/* Open the conveyor belt. */
	cb = ConveyorBeltOpen(rel, DEADTID_FORKNUM, CurrentMemoryContext);

	/* 
	 * Read the conveyor belt bounds and check page validity,
	 * XXX eventually here this function should be called only for assert
	 * checking.
	 */
	ConveyorBeltGetBounds(cb, &oldest_pageno, &next_pageno);
	if (start_pageno < oldest_pageno || start_pageno >= next_pageno)
		return 0;

	/* Loop until we read maxtids. */
	while(ndeadtids < maxtids && start_pageno < next_pageno)
	{
		buffer = ConveyorBeltReadBuffer(cb, start_pageno, BUFFER_LOCK_SHARE, NULL);
		if (BufferIsInvalid(buffer))
			break;

		page = BufferGetPage(buffer);
		count = DTS_PAGE_TID_COUNT(page);

		/*
		 * Don't stop reading in middle of the page, so if remaning tids to
		 * read are less than the tid in this page then stop here.
		 */
		if (maxtids - ndeadtids < count)
			break;

		/* Read data into the deadtid buffer. */
		memcpy(deadtids + ndeadtids, PageGetContents(page),
			   sizeof(ItemPointerData) * count);
		UnlockReleaseBuffer(buffer);

		ndeadtids += count;
		prevpageno = start_pageno;
		start_pageno++;
	}

	*last_pageread = prevpageno;

	return ndeadtids;
}
