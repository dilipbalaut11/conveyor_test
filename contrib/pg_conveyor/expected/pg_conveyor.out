CREATE EXTENSION pg_conveyor;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
 pg_conveyor_init 
------------------
 
(1 row)

SELECT pg_conveyor_insert('test'::regclass::oid, 'test_data');
 pg_conveyor_insert 
--------------------
 
(1 row)

SELECT pg_conveyor_read('test'::regclass::oid, 0);
 pg_conveyor_read 
------------------
 test_data
(1 row)

--CASE1
do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..1000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
-- read from some random blocks
SELECT pg_conveyor_read('test'::regclass::oid, 100);
               pg_conveyor_read                
-----------------------------------------------
 test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa100
(1 row)

SELECT pg_conveyor_read('test'::regclass::oid, 800);
               pg_conveyor_read                
-----------------------------------------------
 test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa800
(1 row)

--CASE2
do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..5000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i+1000;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
SELECT pg_conveyor_read('test'::regclass::oid, 4000);
                pg_conveyor_read                
------------------------------------------------
 test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa4000
(1 row)

SELECT pg_conveyor_read('test'::regclass::oid, 3000);
                pg_conveyor_read                
------------------------------------------------
 test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa3000
(1 row)

--CASE3
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
 pg_conveyor_init 
------------------
 
(1 row)

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..50000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
--CASE4--(vacuum is failing)
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
 pg_conveyor_init 
------------------
 
(1 row)

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..5000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
SELECT pg_conveyor_truncate('test'::regclass::oid, 3000);
 pg_conveyor_truncate 
----------------------
 
(1 row)

SELECT pg_conveyor_vacuum('test'::regclass::oid);
ERROR:  CBM_OBSOLETE_SEGMENT_ENTRIES case is not implemented yet
--CASE5
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
 pg_conveyor_init 
------------------
 
(1 row)

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..50000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
--CASE6 (multi truncate single vacuum)
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
 pg_conveyor_init 
------------------
 
(1 row)

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..1000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
--CASE7
SELECT pg_conveyor_truncate('test'::regclass::oid, 500);
 pg_conveyor_truncate 
----------------------
 
(1 row)

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..1000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
SELECT pg_conveyor_truncate('test'::regclass::oid, 1800);
 pg_conveyor_truncate 
----------------------
 
(1 row)

SELECT pg_conveyor_vacuum('test'::regclass::oid);
 pg_conveyor_vacuum 
--------------------
 
(1 row)

--CASE8
DROP TABLE test;
CREATE TABLE test(a int);
SELECT pg_conveyor_init('test'::regclass::oid, 4);
 pg_conveyor_init 
------------------
 
(1 row)

do $$
<<first_block>>
declare
  i int := 0;
  data varchar;
begin
  for i in 1..5000000 loop
	data := 'test_dataaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' || i;
	PERFORM pg_conveyor_insert('test'::regclass::oid, data);
  end loop;
end first_block $$;
